cmd/storer2elastic/dataprocessor/dataReplayer.go

```diff
@@ -0,0 +1,578 @@
+		if err != nil {
+			return err
+		}
+	}
+
+	defer func() {
+		for _, shardPersisters := range shardPersistersHolder {
+			dr.closePersisters(shardPersisters)
+		}
+	}()
+
+	metachainPersisters, err := dr.preparePersistersHolder(record)
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		dr.closePersisters(metachainPersisters)
+	}()
+
+	epochStartMetaBlock, err := dr.getEpochStartMetaBlock(record, metaHeadersPersisters)
+	if err != nil {
+		return err
+	}
+
+	startingNonce := epochStartMetaBlock.Nonce
+	roundData, err := dr.processMetaBlock(
+		epochStartMetaBlock,
+		dbsInfo,
+		record,
+		metaHeadersPersisters,
+		metachainPersisters,
+		shardPersistersHolder,
+	)
+	if err != nil {
+		return err
+	}
+	if !persistedDataHandler(*roundData) {
+		return ErrRangeIsOver
+	}
+
+	for {
+		startingNonce++
+		metaBlock, err := dr.getMetaBlockForNonce(startingNonce, record, metaHeadersPersisters)
+		if err == nil {
+			roundData, err = dr.processMetaBlock(
+				metaBlock,
+				dbsInfo,
+				record,
+				metaHeadersPersisters,
+				metachainPersisters,
+				shardPersistersHolder,
+			)
+			if err != nil {
+				log.Warn(err.Error())
+				break
+			}
+			if !persistedDataHandler(*roundData) {
+				return ErrRangeIsOver
+			}
+		} else {
+			log.Error(err.Error())
+			break
+		}
+	}
+
+	log.Info("finished indexing all headers from an epoch", "epoch", record.Epoch)
+	return nil
+}
+
+func (dr *dataReplayer) prepareMetaPersistersHolder(record *databasereader.DatabaseInfo) (*metaBlocksPersistersHolder, error) {
+	metaPersHolder := &metaBlocksPersistersHolder{}
+
+	metaBlocksUnit, err := dr.databaseReader.LoadPersister(record, dr.generalConfig.MetaBlockStorage.DB.FilePath)
+	if err != nil {
+		return nil, err
+	}
+	metaPersHolder.metaBlocksPersister = metaBlocksUnit
+
+	headerHashNonceUnit, err := dr.databaseReader.LoadStaticPersister(record, dr.generalConfig.MetaHdrNonceHashStorage.DB.FilePath)
+	if err != nil {
+		return nil, err
+	}
+	metaPersHolder.hdrHashNoncePersister = headerHashNonceUnit
+
+	return metaPersHolder, nil
+}
+
+func (dr *dataReplayer) closeMetaPersisters(persistersHolder *metaBlocksPersistersHolder) {
+	err := persistersHolder.metaBlocksPersister.Close()
+	log.LogIfError(err)
+
+	err = persistersHolder.hdrHashNoncePersister.Close()
+	log.LogIfError(err)
+}
+
+func (dr *dataReplayer) getEpochStartMetaBlock(record *databasereader.DatabaseInfo, metaPersisters *metaBlocksPersistersHolder) (*block.MetaBlock, error) {
+	epochStartIdentifier := core.EpochStartIdentifier(record.Epoch)
+	metaBlockBytes, err := metaPersisters.metaBlocksPersister.Get([]byte(epochStartIdentifier))
+	if err == nil && len(metaBlockBytes) > 0 {
+		return dr.headerMarshalizer.UnmarshalMetaBlock(metaBlockBytes)
+	}
+	log.Warn("epoch start meta block not found", "epoch", record.Epoch)
+
+	// header should be found by the epoch start identifier. if not, try getting the nonce 0 (the genesis block)
+	return dr.getMetaBlockForNonce(0, record, metaPersisters)
```

Comment:
Shouldn't be this an unrecoverable error? What if the the header is really missing?

Reply:
first it will search for `epochStartBlock_X` and if not found, search for nonce 0 in meta hdr hash nonce storer. if still not found, will return error and the program will exit

Reply:
But why this pattern? What if the epoch start block X is missing and you return the block with nonce 0? Does the program will function in a consistent way?

Reply:
refactored. if record's epoch is 0, then search for nonce 0. otherwise, it means that the epoch start meta block was not found and the app stops